File: limits.go
package drpc

// This file holds the configurable delivery buffers (decision Q3) and the
// resource caps (PROTOCOL.md §15).

// DropPolicy selects what a full per-stream rx buffer discards in unreliable
// mode (PROTOCOL.md §4.2).
type DropPolicy int

const (
	// DropNewest discards the arriving frame (the default): the buffered
	// prefix is preserved.
	DropNewest DropPolicy = iota
	// DropOldest discards the oldest buffered frame to admit the newest —
	// freshest-wins, suited to state-sync / sensor streams where the latest
	// reading supersedes older ones.
	DropOldest
)

// rxConfig is a resolved per-stream buffer setting.
type rxConfig struct {
	size   int
	policy DropPolicy
}

const (
	defaultRxBuffer       = 32
	defaultMaxTombs       = 1024
	defaultMaxTombBytes   = 1 << 20
	defaultMaxDeadPeers   = 4
	defaultMaxResets      = 1024
	defaultMaxLiveCalls   = 4096
	defaultRxDropNewest   = DropNewest
	defaultMaxHandlerZero = 0
)

func (c rxConfig) withDefaults() rxConfig {
	if c.size <= 0 {
		c.size = defaultRxBuffer
	}
	return c
}

// Limits bounds the server's per-peer bookkeeping (PROTOCOL.md §15). Zero
// fields keep their defaults. On a Conn only MaxPendingResets applies.
type Limits struct {
	// MaxTombstones caps stored tombstone entries per peer incarnation;
	// oldest are dropped to key-only past it.
	MaxTombstones int
	// MaxTombstoneBytes caps stored terminal-frame payload bytes per peer;
	// oldest stored terminals degrade to key-only past it.
	MaxTombstoneBytes int
	// MaxDeadPeers caps retained finished peer incarnations per transport
	// peer; oldest are evicted (never one with live calls).
	MaxDeadPeers int
	// MaxPendingResets caps the RESET rate-limit / delayed-RESET maps.
	MaxPendingResets int
	// MaxLiveCalls caps concurrently live calls per peer incarnation; an
	// OPEN past it is refused with RESOURCE_EXHAUSTED, bounding the handler
	// goroutines a single peer can spawn.
	MaxLiveCalls int
}

func (l Limits) withDefaults() Limits {
	if l.MaxTombstones <= 0 {
		l.MaxTombstones = defaultMaxTombs
	}
	if l.MaxTombstoneBytes <= 0 {
		l.MaxTombstoneBytes = defaultMaxTombBytes
	}
	if l.MaxDeadPeers <= 0 {
		l.MaxDeadPeers = defaultMaxDeadPeers
	}
	if l.MaxPendingResets <= 0 {
		l.MaxPendingResets = defaultMaxResets
	}
	if l.MaxLiveCalls <= 0 {
		l.MaxLiveCalls = defaultMaxLiveCalls
	}
	return l
}

// WithRxBuffer sets the default per-stream rx buffer size and drop policy for
// every call (PROTOCOL.md §4.2, decision Q3). size <= 0 keeps the default 32.
func WithRxBuffer(size int, policy DropPolicy) interface {
	ConnOption
	ServerOption
} {
	return limitsOption{rx: &rxConfig{size: size, policy: policy}}
}

// WithMethodRxBuffer overrides the rx buffer for one full method name
// (e.g. "/echo.EchoService/Live") on the server — a high-rate sensor stream
// can carry a deeper buffer and DropOldest while control RPCs keep the
// default. Most-specific wins: method override, else the WithRxBuffer
// default, else the built-in default.
func WithMethodRxBuffer(fullMethod string, size int, policy DropPolicy) ServerOption {
	return serverOptionFunc(func(o *serverOption) {
		if o.methodRx == nil {
			o.methodRx = map[string]rxConfig{}
		}
		o.methodRx[fullMethod] = rxConfig{size: size, policy: policy}
	})
}

// WithLimits sets the resource caps (PROTOCOL.md §15). Zero fields keep
// defaults.
func WithLimits(l Limits) interface {
	ConnOption
	ServerOption
} {
	return limitsOption{limits: &l}
}

type limitsOption struct {
	rx     *rxConfig
	limits *Limits
}

func (o limitsOption) apply(c *connOption) {
	if o.rx != nil {
		c.rx = *o.rx
	}
	if o.limits != nil {
		c.limits = *o.limits
	}
}

func (o limitsOption) applyServer(s *serverOption) {
	if o.rx != nil {
		s.rx = *o.rx
	}
	if o.limits != nil {
		s.limits = *o.limits
	}
}

File: ts/src/limits.ts
// Configurable delivery buffers (PROTOCOL.md §4.2) and resource caps (§15).

// DropPolicy selects what a full per-stream rx buffer discards in unreliable
// mode (PROTOCOL.md §4.2). Reliable mode never drops: delivery blocks instead.
export enum DropPolicy {
  // Discard the arriving frame (the default): the buffered prefix is
  // preserved.
  Newest,
  // Discard the oldest buffered frame to admit the newest — freshest-wins,
  // suited to state-sync / sensor streams.
  Oldest,
}

export interface RxBufferConfig {
  size?: number
  policy?: DropPolicy
}

export interface ResolvedRxConfig {
  size: number
  policy: DropPolicy
}

export function resolveRxConfig(c: RxBufferConfig = {}): ResolvedRxConfig {
  const size = c.size !== undefined && c.size > 0 ? c.size : 32
  return { size, policy: c.policy ?? DropPolicy.Newest }
}

// Limits bounds the endpoint's bookkeeping (PROTOCOL.md §15). Absent fields
// keep their defaults. On a Conn only maxPendingResets and maxPeerWindow
// apply.
export interface Limits {
  // Caps stored tombstone entries per peer incarnation. Past it the lowest
  // sid is evicted and the container's floor rises: evicted sids keep
  // key-only semantics (deduped, replay lost) at zero memory.
  maxTombstones?: number
  // Caps stored terminal-frame payload bytes per peer incarnation; oldest
  // stored terminals degrade to key-only past it.
  maxTombstoneBytes?: number
  // Caps retained finished peer incarnations per transport peer; oldest are
  // evicted (never one with live calls).
  maxDeadPeers?: number
  // Caps the RESET rate-limit / delayed-RESET / reply-budget maps.
  maxPendingResets?: number
  // Caps concurrently live calls per transport peer, counted across client
  // epochs; an OPEN past it is refused with RESOURCE_EXHAUSTED.
  maxLiveCalls?: number
  // Caps, per transport peer, the control replies the server volunteers
  // within one RTI — tombstone/creation-ack replays and RESETs — on top of
  // the per-object 1/RTI limits (anti-amplification).
  maxRepliesPerRTI?: number
  // Caps, in messages, what one transport peer may have buffered here across
  // all of its calls and client epochs — the connection flow-control window
  // (§4.2.1, reliable mode only); on a Conn it bounds the one peer the Conn
  // talks to. It is advertised as Frame.connWindow — on every OPEN by a
  // Conn, on every H and T by a Server — and the peer honours the first
  // advertisement it hears. Values below W_CONN (1024) are raised to it: a
  // client streams on the W_CONN assumption until the server's first H or T
  // arrives, so a receiver holding less would be overrun by a conforming
  // sender — the same reason the rx buffer is floored at W_INIT. Past it,
  // the frame that overruns fails its own call with INTERNAL, never the
  // peer. Capped at 2^32 − 1, the wire's uint32 (§5): Infinity means that
  // much, not "off".
  maxPeerWindow?: number
}

// DEFAULT_MAX_PEER_WINDOW is the default and the floor of maxPeerWindow: it
// equals W_CONN (util.ts), the client's assumption — spelled out here rather
// than imported, since util.ts already imports this module.
const DEFAULT_MAX_PEER_WINDOW = 1024

// MAX_PEER_WINDOW is the cap of maxPeerWindow: `conn_window` is a uint32 on
// the wire (§5), and the window is what the advertisement puts there
// (§4.2.1) and what the grant rule measures against. Go's field is an int
// cast to uint32 at the ledger, so it can never hold more; in JS a number
// can, and Infinity is the natural spelling of "unlimited" — unclamped it
// would reach the ledger, where no batched or starvation grant ever fires
// against it, and the wire, whose uint32 varint cannot carry it. Clamped,
// it is a window no peer can fill, and the advertisement says exactly that.
const MAX_PEER_WINDOW = 0xffff_ffff

export interface ResolvedLimits {
  maxTombstones: number
  maxTombstoneBytes: number
  maxDeadPeers: number
  maxPendingResets: number
  maxLiveCalls: number
  maxRepliesPerRTI: number
  maxPeerWindow: number
}

export function resolveLimits(l: Limits = {}): ResolvedLimits {
  const pos = (v: number | undefined, d: number) => (v !== undefined && v > 0 ? v : d)
  return {
    maxTombstones: pos(l.maxTombstones, 1024),
    maxTombstoneBytes: pos(l.maxTombstoneBytes, 1 << 20),
    maxDeadPeers: pos(l.maxDeadPeers, 4),
    maxPendingResets: pos(l.maxPendingResets, 1024),
    maxLiveCalls: pos(l.maxLiveCalls, 4096),
    maxRepliesPerRTI: pos(l.maxRepliesPerRTI, 64),
    // Floored, not just defaulted (the client's assumption), capped at the
    // wire's uint32, and whole: a message count. Infinity clamps to the cap;
    // NaN, like any non-positive value, keeps the default.
    maxPeerWindow: Math.min(Math.max(Math.floor(pos(l.maxPeerWindow, DEFAULT_MAX_PEER_WINDOW)), DEFAULT_MAX_PEER_WINDOW), MAX_PEER_WINDOW),
  }
}
